File: src/services/dashboard/websocket.py
"""
CobaltGraph WebSocket Event Handlers
Room-based real-time updates for network monitoring dashboard
"""

from flask_socketio import emit, join_room, leave_room, rooms
from flask import request
from datetime import datetime
import logging

logger = logging.getLogger(__name__)


def register_socketio_events(socketio, database):
    """
    Register all SocketIO event handlers

    Args:
        socketio: SocketIO instance
        database: Database instance
    """

    @socketio.on('connect')
    def handle_connect():
        """Client connected"""
        client_id = request.sid
        logger.info(f"🔌 Client connected: {client_id}")

        emit('connection_established', {
            'status': 'connected',
            'client_id': client_id,
            'timestamp': str(datetime.now())
        })

    @socketio.on('disconnect')
    def handle_disconnect():
        """Client disconnected"""
        client_id = request.sid
        logger.info(f"🔌 Client disconnected: {client_id}")

    @socketio.on('subscribe_device_list')
    def handle_subscribe_device_list():
        """Subscribe to device list updates"""
        client_id = request.sid
        join_room('device_list')

        logger.info(f"📋 Client {client_id} subscribed to device_list")

        # Send current device count
        try:
            # Get device count from database
            devices = database.get_devices(limit=10000)
            device_count = len(devices)
            emit('device_count_update', {'count': device_count})
        except Exception as e:
            logger.error(f"Error getting device count: {e}")

    @socketio.on('unsubscribe_device_list')
    def handle_unsubscribe_device_list():
        """Unsubscribe from device list updates"""
        client_id = request.sid
        leave_room('device_list')
        logger.info(f"📋 Client {client_id} unsubscribed from device_list")

    @socketio.on('subscribe_device')
    def handle_subscribe_device(data):
        """
        Subscribe to specific device updates

        Args:
            data: {'mac_address': 'AA:BB:CC:DD:EE:FF'}
        """
        client_id = request.sid
        mac_address = data.get('mac_address')

        if not mac_address:
            emit('error', {'message': 'mac_address required'})
            return

        room = f'device_{mac_address}'
        join_room(room)

        logger.info(f"📱 Client {client_id} subscribed to device {mac_address}")

        # Send current device data
        try:
            device = database.get_device_by_mac(mac_address)
            if device:
                emit('device_data', device)
            else:
                emit('error', {'message': 'Device not found'})
        except Exception as e:
            logger.error(f"Error getting device {mac_address}: {e}")
            emit('error', {'message': str(e)})

    @socketio.on('unsubscribe_device')
    def handle_unsubscribe_device(data):
        """Unsubscribe from specific device updates"""
        client_id = request.sid
        mac_address = data.get('mac_address')

        if mac_address:
            room = f'device_{mac_address}'
            leave_room(room)
            logger.info(f"📱 Client {client_id} unsubscribed from device {mac_address}")

    @socketio.on('request_device_list')
    def handle_request_device_list(data):
        """
        Request device list (with filters)

        Args:
            data: {
                'status': 'active' | 'offline' | 'idle' | None,
                'limit': 100
            }
        """
        try:
            status = data.get('status') if data else None
            limit = data.get('limit', 100) if data else 100

            devices = database.get_devices(status=status, limit=limit)

            emit('device_list_data', {
                'devices': devices,
                'total': len(devices),
                'timestamp': str(datetime.now())
            })

        except Exception as e:
            logger.error(f"Error getting device list: {e}")
            emit('error', {'message': str(e)})

    # ========================================================================
    # Server-side event emitters (called from Device Discovery Service)
    # ========================================================================

    def emit_device_discovered(device_data):
        """
        Emit device_discovered event to device_list room

        Args:
            device_data: Device dictionary
        """
        socketio.emit('device_discovered', device_data, room='device_list')
        logger.debug(f"📡 Emitted device_discovered: {device_data.get('mac_address')}")

    def emit_device_updated(device_data):
        """
        Emit device_updated event to both device_list and device-specific room

        Args:
            device_data: Device dictionary
        """
        mac = device_data.get('mac_address')

        # Emit to device list
        socketio.emit('device_updated', device_data, room='device_list')

        # Emit to device-specific room
        socketio.emit('device_updated', device_data, room=f'device_{mac}')

        logger.debug(f"📡 Emitted device_updated: {mac}")

    def emit_device_status_changed(mac_address, old_status, new_status):
        """
        Emit device_status_changed event

        Args:
            mac_address: Device MAC
            old_status: Previous status
            new_status: New status
        """
        data = {
            'mac_address': mac_address,
            'old_status': old_status,
            'new_status': new_status,
            'timestamp': str(datetime.now())
        }

        socketio.emit('device_status_changed', data, room='device_list')
        socketio.emit('device_status_changed', data, room=f'device_{mac_address}')

        logger.debug(f"📡 Emitted status change: {mac_address} {old_status} → {new_status}")

    def emit_connection_added(connection_data):
        """
        Emit connection_added event to device-specific room

        Args:
            connection_data: Connection dictionary
        """
        mac = connection_data.get('src_mac')
        if mac:
            socketio.emit('connection_added', connection_data, room=f'device_{mac}')
            logger.debug(f"📡 Emitted connection_added for device {mac}")

    # Store emitter functions on socketio object for external access
    socketio.emit_device_discovered = emit_device_discovered
    socketio.emit_device_updated = emit_device_updated
    socketio.emit_device_status_changed = emit_device_status_changed
    socketio.emit_connection_added = emit_connection_added

    logger.info("✅ WebSocket events registered")

File: src/services/dashboard/static/js/websocket.js
/**
 * CobaltGraph WebSocket Client
 * Room-based real-time updates for network monitoring
 */

class SuaronWebSocket {
    constructor() {
        this.socket = null;
        this.connected = false;
        this.reconnectAttempts = 0;
        this.maxReconnectAttempts = 5;
        this.reconnectDelay = 2000; // ms

        this.eventHandlers = {};
        this.rooms = new Set();
    }

    /**
     * Connect to WebSocket server
     */
    connect() {
        console.log('🔌 Connecting to WebSocket server...');

        this.socket = io({
            transports: ['websocket', 'polling'],
            reconnection: true,
            reconnectionDelay: this.reconnectDelay,
            reconnectionAttempts: this.maxReconnectAttempts
        });

        this._registerCoreEvents();

        return this;
    }

    /**
     * Register core socket events
     */
    _registerCoreEvents() {
        this.socket.on('connect', () => {
            console.log('✅ WebSocket connected');
            this.connected = true;
            this.reconnectAttempts = 0;
            this._updateConnectionStatus('connected');

            // Resubscribe to rooms after reconnect
            this._resubscribeRooms();
        });

        this.socket.on('disconnect', (reason) => {
            console.warn('❌ WebSocket disconnected:', reason);
            this.connected = false;
            this._updateConnectionStatus('disconnected');
        });

        this.socket.on('connect_error', (error) => {
            console.error('⚠️ WebSocket connection error:', error);
            this._updateConnectionStatus('error');
        });

        this.socket.on('error', (error) => {
            console.error('⚠️ WebSocket error:', error);
        });

        this.socket.on('connection_established', (data) => {
            console.log('🎉 Connection established:', data);
        });
    }

    /**
     * Update connection status indicator in UI
     */
    _updateConnectionStatus(status) {
        const statusEl = document.getElementById('connection-status');
        if (!statusEl) return;

        const icon = statusEl.querySelector('i');
        const text = statusEl.querySelector('span');

        switch (status) {
            case 'connected':
                icon.className = 'bi bi-circle-fill text-success';
                text.textContent = 'Connected';
                break;
            case 'disconnected':
                icon.className = 'bi bi-circle-fill text-warning';
                text.textContent = 'Disconnected';
                break;
            case 'error':
                icon.className = 'bi bi-circle-fill text-danger';
                text.textContent = 'Connection Error';
                break;
        }
    }

    /**
     * Resubscribe to all rooms after reconnect
     */
    _resubscribeRooms() {
        for (const room of this.rooms) {
            if (room === 'device_list') {
                this.subscribeDeviceList();
            } else if (room.startsWith('device_')) {
                const mac = room.replace('device_', '');
                this.subscribeDevice(mac);
            }
        }
    }

    /**
     * Subscribe to device list updates
     */
    subscribeDeviceList() {
        console.log('📋 Subscribing to device_list room');
        this.socket.emit('subscribe_device_list');
        this.rooms.add('device_list');
    }

    /**
     * Unsubscribe from device list updates
     */
    unsubscribeDeviceList() {
        console.log('📋 Unsubscribing from device_list room');
        this.socket.emit('unsubscribe_device_list');
        this.rooms.delete('device_list');
    }

    /**
     * Subscribe to specific device updates
     * @param {string} macAddress - Device MAC address
     */
    subscribeDevice(macAddress) {
        console.log(`📱 Subscribing to device ${macAddress}`);
        this.socket.emit('subscribe_device', { mac_address: macAddress });
        this.rooms.add(`device_${macAddress}`);
    }

    /**
     * Unsubscribe from specific device
     * @param {string} macAddress - Device MAC address
     */
    unsubscribeDevice(macAddress) {
        console.log(`📱 Unsubscribing from device ${macAddress}`);
        this.socket.emit('unsubscribe_device', { mac_address: macAddress });
        this.rooms.delete(`device_${macAddress}`);
    }

    /**
     * Request device list
     * @param {Object} options - Filter options
     */
    requestDeviceList(options = {}) {
        this.socket.emit('request_device_list', options);
    }

    /**
     * Register event handler
     * @param {string} event - Event name
     * @param {Function} handler - Handler function
     */
    on(event, handler) {
        if (!this.eventHandlers[event]) {
            this.eventHandlers[event] = [];
            this.socket.on(event, (data) => {
                this.eventHandlers[event].forEach(h => h(data));
            });
        }
        this.eventHandlers[event].push(handler);
    }

    /**
     * Disconnect from server
     */
    disconnect() {
        if (this.socket) {
            this.socket.disconnect();
        }
    }
}

// Global instance
window.suaronWS = new SuaronWebSocket();
